### PR TITLE
check the target host, port and scheme for None values after URL parsing

### PR DESCRIPTION
**Pull request by ybadaoui-ostorlab**

### CODE REVIEW

File: tests/targets_preparer_test.py [hunk 1]
```diff
@@ -65,3 +66,11 @@ def testPrepareTargets_whenIPv6AssetReachCIDRLimit_raiseValueError(
 
     with pytest.raises(ValueError, match="Subnet mask below 112 is not supported."):
         assert any(targets)
+
+
+def testPrepareTargets_whenBadMessage_returnEmptyTargets(
```
Comment (elyousfi5) [inline tests/targets_preparer_test.py:71]:
```suggestion
def testPrepareTargets_whenUrlCantBeParsed_returnEmptyTargets(
```

File: agent/targets_preparer.py [hunk 2]
```diff
@@ -99,7 +99,8 @@ def prepare_targets(message: m.Message) -> Generator[definitions.Target, None, N
         parsed_url = urlparse(url)
         host = parsed_url.hostname
         port = parsed_url.port or _get_port(message, parsed_url.scheme)
-        scheme = parsed_url.scheme
-        yield definitions.Target(host=host, port=port, scheme=scheme)
+        scheme = parsed_url.scheme or "https"
+        if host is not None and port is not None and scheme is not None:
+            yield (definitions.Target(host=host, port=port, scheme=scheme))
     else:
         raise NotImplementedError
```
Comment (nmasdoufi-ol) [inline agent/targets_preparer.py:106]:
```suggestion
        logger.warning(f"Unsupported target type in message. Message data: {message.data}")
        raise NotImplementedError()
```

File: agent/targets_preparer.py [hunk 3]
```diff
@@ -100,6 +100,7 @@ def prepare_targets(message: m.Message) -> Generator[definitions.Target, None, N
         host = parsed_url.hostname
         port = parsed_url.port or _get_port(message, parsed_url.scheme)
         scheme = parsed_url.scheme
-        yield definitions.Target(host=host, port=port, scheme=scheme)
+        if host is not None and port is not None and scheme is not None:
+            yield (definitions.Target(host=host, port=port, scheme=scheme))
```
[the remark this answers is not in the document]
Comment (ybadaoui-ostorlab) [reply agent/targets_preparer.py:107]: fixed

File: tests/targets_preparer_test.py (hunk 1 above)
Comment (ybadaoui-ostorlab) [reply tests/targets_preparer_test.py]: fixed

File: tests/targets_preparer_test.py [hunk 4]
```diff
@@ -65,3 +66,12 @@ def testPrepareTargets_whenIPv6AssetReachCIDRLimit_raiseValueError(
...
         assert any(targets)
+
+
+def testPrepareTargets_whenUrlCantBeParsed_returnEmptyTargets(
+    scan_bad_message: message.Message,
+) -> None:
+    targets = targets_preparer.prepare_targets(scan_bad_message)
+
+    with pytest.raises(NotImplementedError):
+        assert any(targets) is False
```
Comment (rabsondev) [inline tests/targets_preparer_test.py:77]: Can you make `any(targets)` more explicit? It's hard to tell what it's checking for.

File: agent/targets_preparer.py [hunk 5]
```diff
@@ -99,7 +102,13 @@ def prepare_targets(message: m.Message) -> Generator[definitions.Target, None, N
...
         port = parsed_url.port or _get_port(message, parsed_url.scheme)
-        scheme = parsed_url.scheme
-        yield definitions.Target(host=host, port=port, scheme=scheme)
+        scheme = parsed_url.scheme or "https"
+        if host is not None and port is not None and scheme is not None:
+            yield (definitions.Target(host=host, port=port, scheme=scheme))
+        else:
+            logger.warning(
+                "Incomplete target configuration: host, port, and scheme must all be provided."
+            )
```
Comment (mohsinenar) [inline agent/targets_preparer.py:111]: Log also the value of the target.

Comment (ybadaoui-ostorlab) [reply agent/targets_preparer.py]: done

**Pull request merged**